lib/services/elasticsearch.js

```diff
@@ -153,665 +173,682 @@ class ElasticSearch extends Service {
+   *
+   * @returns {Promise} { }
+   */
+  createCollection (index, collection, mappings = {}) {
+    const esRequest = {
+      index: this._getESIndex(index, collection),
+      body: _.merge(mappings, this.config.commonMapping)
+    };
 
-        if (esRequestBulk.body.length === 0) {
-          return Bluebird.resolve({ids: []});
+    return this.kuzzle.indexCache.exists(index)
+      .then(indexExists => {
+        if (! indexExists) {
+          this.throw('index_does_not_exist');
         }
 
-        return this.client.bulk(esRequestBulk)
-          .then(() => this.refreshIndexIfNeeded(esRequestBulk, {ids}))
-          .catch(error => Bluebird.reject(
-            this.esWrapper.formatESError(error))
-          );
+        return this.kuzzle.indexCache.exists(index, collection);
+      })
+      .then(exists => {
+        if (exists) {
+          return this.updateMapping(index, collection, mappings);
+        }
+
+        this._checkMappings(esRequest.body.mappings);
+
+        return this.client.indices.create(esRequest)
+          .then(() => {
+            const apiResponse = {};
+
+            return apiResponse;
+          })
+          .catch(error => this.esWrapper.reject(error));
       });
   }
 
   /**
-   * Delete all document that match the given filter from the trash
-   * @param {Request} request
-   * @returns {Promise} resolve the list of deleted ids
+   * Add a mapping definition to a specific type
+   *
+   * @param {string} index - Index name
+   * @param {string} collection - Collection name
+   * @param {object} mappings - Collection mappings in ES format
+   *
+   * @return {Promise} { }
    */
-  deleteByQueryFromTrash(request) {
-    const
-      esRequestSearch = initESRequest(request, ['from', 'size', 'scroll']),
-      esRequestBulk = initESRequest(request, ['refresh']);
-
-    assertWellFormedRefresh(esRequestBulk);
+  updateMapping (index, collection, mappings) {
+    const esRequest = {
+      index: this._getESIndex(index, collection)
+    };
 
-    esRequestSearch.body = request.input.body;
-    esRequestSearch.scroll = '30s';
+    return this.getMapping(index, collection, true)
+      .then(collectionMappings => {
+        this._checkMappings(collectionMappings);
 
-    if (esRequestSearch.body.query === null) {
-      return Bluebird.reject(this.getError('document_id_cannot_be_null'));
-    }
+        esRequest.body = _.merge(mappings, collectionMappings);
 
-    return getAllIdsFromQuery(this.client, esRequestSearch)
-      .then(ids => {
-        return new Bluebird((resolve, reject) => {
-          esRequestBulk.body = ids.map(id => ({delete:
-            {_index: esRequestBulk.index,
-              _type: esRequestBulk.type,
-              _id: id}
-          }));
-
-          if (esRequestBulk.body.length === 0) {
-            return resolve({ids: []});
-          }
+        return this.client.indices.putMapping(esRequest)
+          .then(() => {
+            const apiResponse = {};
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Immediately return this expression instead of assigning it to the temporary variable "apiResponse". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1488)